Graphics/Vulkan/VulkanGraphicsBuffer.cpp

```diff
@@ -0,0 +1,153 @@
+                                               VkBuffer vulkanBuffer)
+        : VulkanGraphicsResource(graphicsDevice, allocator, cpuAccess, allocation, allocationInfo),
+          _vulkanBuffer(vulkanBuffer),
+          _subAllocations(0),
+          _cpuAccess(cpuAccess),
+          _kind(kind)
+    {}
+
+    VulkanGraphicsBuffer::~VulkanGraphicsBuffer() noexcept
+    {
+        auto allocator = GetAllocator()->GetVmaAllocator();
+        auto allocation = GetAllocation();
+
+        assert_message("Attempted to destroy a VkBuffer containing mapped regions.", _subAllocations != 0);
+
+        vmaDestroyBuffer(allocator, GetVulkanBuffer(), allocation);
+    }
+
+    Utilities::Misc::Span<uint8_t> VulkanGraphicsBuffer::MapBytes(size_t rangeOffset, size_t rangeLength)
+    {
+        size_t sizeOfBuffer = GetAllocationInfo().size;
+        size_t rangeValidationValue = sizeOfBuffer - rangeOffset;
+
+        if (rangeValidationValue < rangeLength)
+        {
+            throw Exceptions::InvalidOperationException(
+                "Attempted to map a subrange of a VkBuffer that exceeded the VkBuffer's size.");
+        }
+
+        void* data = nullptr;
+        VkResult result =
+            vmaMapMemory(GetAllocator()->GetVmaAllocator(), GetAllocation(), &data);
+
+        if (result != VK_SUCCESS)
+        {
+            // TODO: Make this a real exception
+            throw std::runtime_error("Failed to map Vulkan memory to the CPU. Reason: " + std::to_string(result));
+        }
+
+        _subAllocations++;
+
+        return Utilities::Misc::Span<uint8_t>(reinterpret_cast<uint8_t*>(data) + rangeOffset, rangeLength);
+    }
+
+    Utilities::Misc::Span<const uint8_t> VulkanGraphicsBuffer::MapBytesForRead(size_t rangeOffset, size_t rangeLength)
+    {
+        size_t sizeOfBuffer = GetAllocationInfo().size;
+        size_t rangeValidationValue = sizeOfBuffer - rangeOffset;
+
+        if (rangeValidationValue < rangeLength)
+        {
+            throw Exceptions::InvalidOperationException(
+                "Attempted to map a subrange of a VkBuffer that exceeded the VkBuffer's size.");
+        }
+
+        VmaAllocator allocator = GetAllocator()->GetVmaAllocator();
+        VmaAllocation allocation = GetAllocation();
+
+        void* data = nullptr;
+        VkResult mapResult = vmaMapMemory(allocator, allocation, &data);
+
+        if (mapResult != VK_SUCCESS)
+        {
+            // TODO: Make this a real exception
+            throw std::runtime_error("Failed to map Vulkan memory to the CPU. Reason: " + std::to_string(mapResult));
+        }
+
+        VkResult invalidateResult = vmaInvalidateAllocation(allocator, allocation, rangeOffset, rangeLength);
+
+        if (invalidateResult != VK_SUCCESS)
+        {
+            // TODO: Make this a real exception
+            throw std::runtime_error("Failed to invalidate mapped memory. Reason: " + std::to_string(mapResult));
```

Comment:
Same here, unless we don't care at the moment. Exceptions don't have feelings...

Reply:
So I have attempted to do this, however I think this is part of a broader conversation we need to have about asserts vs exceptions in some parts of the graphics API, and what we'd like to do, so I'm going to leave this alone for now.